prov/verbs/src/fi_verbs.c

```diff
@@ -131,6 +133,68 @@ int fi_ibv_create_ep(const char *node, const char *service,
 	return ret;
 }
 
+#define VERBS_SIGNAL_SEND(ep) \
+	(atomic_get(&ep->unsignaled_send_cnt) >= VERBS_SEND_SIGNAL_THRESH(ep) && \
+	 !atomic_get(&ep->comp_pending))
+
+static int fi_ibv_process_unsignaled_send(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr)
+{
+	struct fi_ibv_msg_epe *epe;
+	struct fi_ibv_wce *wce;
+	struct ibv_wc wc;
+	int ret = 0;
+
+	if (VERBS_SIGNAL_SEND(ep)) {
+		fastlock_acquire(&ep->scq->ep_list_lock);
+		if (VERBS_SIGNAL_SEND(ep)) {
+			wr->send_flags |= IBV_SEND_SIGNALED;
+			wr->wr_id = ep->ep_id;
+			epe = util_buf_alloc(ep->scq->epe_pool);
+			if (!epe) {
+				fastlock_release(&ep->scq->ep_list_lock);
+				return -FI_ENOMEM;
+			}
+			epe->ep = ep;
+			slist_insert_tail(&epe->entry, &ep->scq->ep_list);
+			atomic_inc(&ep->comp_pending);
+			fastlock_release(&ep->scq->ep_list_lock);
+			return 0;
+		}
+		fastlock_release(&ep->scq->ep_list_lock);
+	}
+
+	atomic_inc(&ep->unsignaled_send_cnt);
+
+	if (atomic_get(&ep->unsignaled_send_cnt) >= VERBS_SEND_COMP_THRESH(ep)) {
+		fastlock_acquire(&ep->scq->lock);
+		while (atomic_get(&ep->comp_pending) > 0) {
+			ret = fi_ibv_poll_cq(ep->scq, &wc);
+			if (ret < 0) {
+				FI_WARN(&fi_ibv_prov, FI_LOG_EP_DATA,
+					"Failed to read completion for signaled send\n");
+				return ret;
+			} else if (ret > 0) {
+				wce = util_buf_alloc(ep->scq->wce_pool);
```

Comment:
I would look at re-arranging this code to call util_buf_alloc before polling on the CQ.  This avoids the memcpy and allows you to abort if util_buf_alloc fails.